### ISSUE TITLE
Pre-trained Model Predicting All Images as Label 4 - Is This Correct?

### ISSUE THREAD

**Issue by HeeminYang** [JLREx/PAtt-Lite#6]
Hello,

I've been experimenting with the pre-trained model you've uploaded, and I've noticed an issue. Regardless of the input, the model consistently predicts every image to be label 4. This behavior leads me to question whether the correct model has been uploaded.

Could you please confirm if this is the expected behavior of the model? If not, would it be possible to check the uploaded model file and ensure the correct version is available for download?

Thank you for your assistance.

**Comment by ipa-anm-sy** [comment]
I am also having the same issue. The model consistently predicts every image to be label 4. Can you please fix this issue?

**Comment by M-U-X** [comment]
how did you guys get it to work, what inference code did you use? @HeeminYang @ipa-anm-sy

**Comment by ipa-anm-sy** [comment]
I am not using this model for training or evaluation. I am just using it for getting predictions. I used the provided pre-trained model, and used the command
''model=tf.keras.models.load_model(model_path, compile=False)'' , before this I am using face detection algorithm, and cropping the face and resizing the images to (224,224) and image transforms, and I am extracting labels from the output of the model.

**Comment by Chen1fly** [comment]
Hello, could you please tell me where did you get the code?

**Comment by MeneerAbel** [comment]
Sorry but that is just the standard code to run any keras model ... If you plan to work with pre trained models it is still good to do a bit of research yourself ... one google search could have provided you with this. If it is a face detection algorithm you are loooking for, there is again many out there, for example deepface: [https://github.com/serengil/deepface/tree/master](https://github.com/serengil/deepface/tree/master) Let's try to stay on topic. This issue is about the fact that the pre-trained model always predicts label 4, an issue that I would still like to have a reply on by the developers.

**Comment by ipa-anm-sy** [comment]
Can anyone reply on the main question in this issue? The model is only predicting label 4.

**Comment by JLREx** [comment]
Hi, we have just uploaded the training notebook. The evaluation code is at the end of the notebook. Still, it is recommended to train on your database before evaluating the model for optimal performance. Based on our evaluation, the cross-database results are not quite perfect.

**Comment by HeeminYang** [comment]
@JLREx thanks for the update. But this issue is for your model's convergence, and the notebook that you upload has not include any result about the model performance which reach the [SOTA on FER+](https://paperswithcode.com/sota/facial-expression-recognition-on-fer-1). The model which update on April 6 still predicts all label as 4. Your response is not unsuitable for this issue.

**Comment by JLREx** [comment]
> @JLREx thanks for the update. But this issue is for your model's convergence, and the notebook that you upload has not include any result about the model performance which reach the [SOTA on FER+](https://paperswithcode.com/sota/facial-expression-recognition-on-fer-1). The model which update on April 6 still predicts all label as 4. Your response is not unsuitable for this issue.

Hi, to clarify your issues:
1. As mentioned, cross-database results are not great. I'm not sure what data were you evaluating on previously, but this could be the reason the pre-trained model only predicted label 4. For FERPlus, it is the same cause, model trained on RAF-DB is not supposed to reach SOTA on FERPlus. The model was uploaded as proof for evaluation on the RAF-DB testing set that we achieved the performance reported in the paper. 
2. The notebook uploaded recently has nothing to do with the REMOVED pre-trained model. It is for your training on your dataset (FER2013, FERPlus, or your dataset) and to evaluate the trained model afterward. Hence, if you wanted to get the reported FERPlus results, you will have to download the dataset and train with the notebook/the model architecture in the notebook, before evaluating on FERPlus. 

Hope this helps.

**Comment by HeeminYang** [comment]
@JLREx I now understand. The model you uploaded is related to RAF-DB and has nothing to do with FER+, right? However, on the paperswithcode website, your PAtt-Lite is registered as achieving SOTA on FER+ with an accuracy of 95.550. It seems that many users are raising questions about the low performance of the model because of this. It seems that a clear notice is necessary, and do you have any plans to release the model that achieved SOTA registered on the paperswithcode website?

**Comment by JLREx** [comment]
@HeeminYang Yes, the model trained on RAF-DB has nothing to do with FERPlus. The results registered on paperswithcode is a model trained and evaluated on FERPlus. For now, we do not have any plan to release the model as it could potentially cause further confusion. You may train your model with the notebook shared in this repository. From our runs on Kaggle, a training on FERPlus should be around or less than 2 hours.

**Comment by MyungBeomHer** [comment]
@JLREx  The model inputs 2D tensor to "tf.keras.layers.Attention".
More specifically, the model use many conv2d layer, and then output is 4D tensor [B,C,H,W], but tf.keras.layers.GlobalAveragePooling2D's input is 4D tensor and output is 2D tensor.
So input of attention layer is 2D tensor. Finally, attention layer needs only 3D tensor[B,N,E] where B=Batch, N=number of patches, E=Embedding dimension. But this model input 2D tensor[B,E] to attention layer. Thus, Attention layer think this input has only one sample, not batch. So they convert [B,E] to [1,B,E] and run the attention operation (dot product). Therefore, when this model inference, they are not dot product about only one sample, dot product among the other samples in validation dataset.  If the different label validation data within a batch exists during inference, their accuracy decreases rapidly.(originally, same label is clustered in validation dataset.ex) label1 ... label1, label2,...label2,... label n ,... , label n) Therefore, it cannot be said that the performance is correct,

**Comment by cyinen** [comment]
@heomyeongboem Hello！
May I ask where you downloaded the pre-trained model? Can anyone give a link?

**Comment by MyungBeomHer** [comment]
@cyinen Hello.
I don't have the pre-trained model. But when I see the patte-lite code, they use conv2d [B,C,H,W] -> GAP [B,C]-> linear [B, C'] -> attention [B,C'] -> linear for classification. So I thought their code have problem. 

tf/keras/layers/GlobalAveragePooling2D
[https://www.tensorflow.org/api_docs/python/tf/keras/layers/GlobalAveragePooling2D](url)

tf/keras/layers/Attention
[https://www.tensorflow.org/api_docs/python/tf/keras/layers/Attention](url)

**Comment by LokmaneZ** [comment]
Hello, I am looking for the pretrained model, does anyone have link to get it? 
Thank you in advance

**Comment by JLREx** [comment]
> @JLREx The model inputs 2D tensor to "tf.keras.layers.Attention". More specifically, the model use many conv2d layer, and then output is 4D tensor [B,C,H,W], but tf.keras.layers.GlobalAveragePooling2D's input is 4D tensor and output is 2D tensor. So input of attention layer is 2D tensor. Finally, attention layer needs only 3D tensor[B,N,E] where B=Batch, N=number of patches, E=Embedding dimension. But this model input 2D tensor[B,E] to attention layer. Thus, Attention layer think this input has only one sample, not batch. So they convert [B,E] to [1,B,E] and run the attention operation (dot product). Therefore, when this model inference, they are not dot product about only one sample, dot product among the other samples in validation dataset. If the different label validation data within a batch exists during inference, their accuracy decreases rapidly.(originally, same label is clustered in validation dataset.ex) label1 ... label1, label2,...label2,... label n ,... , label n) Therefore, it cannot be said that the performance is correct,

Hi. Thanks a lot for pointing it out. We overlooked on how TensorFlow treats and assumes its input. 
We are working on it and will try to update the new model and codes here in the coming weeks.

**Comment by userguazi** [comment]
where can i find pytorch code

**Comment by userguazi** [comment]
您好，我正在寻找预训练模型，怎么才能找到啊

**Comment by JLREx** [comment]
@cyinen @LokmaneZ @userguazi 
Hi, the pretrained model is removed due to the cross-database performance as mentioned in earlier replies. 
You should train your model with the notebook provided instead.

**Comment by KrystianZielinski** [comment]
Hey. I am trying to evaluate Patt-lite using 5-fold CV -  so I trained the Patt-lite model on FER+ on my own folds (different train/val/test sets with 80/10/10 % ratios). What's weird to me is that in the first fold i got around 80% accuracy on train set, around 72% on validation set, but on test set I got... ~95%. And then I train the model again on 2nd fold - I get around same accuracies on train/validation sets, but this time accuracy on test set is only ~33%. So I wanted to ask if you got similar results? And what could be the cause that one time the test accuracy is much higher than on train/val sets and the next time its way lower? Just model not being stable?

**Comment by KrystianZielinski** [comment]
> @JLREx The model inputs 2D tensor to "tf.keras.layers.Attention". More specifically, the model use many conv2d layer, and then output is 4D tensor [B,C,H,W], but tf.keras.layers.GlobalAveragePooling2D's input is 4D tensor and output is 2D tensor. So input of attention layer is 2D tensor. Finally, attention layer needs only 3D tensor[B,N,E] where B=Batch, N=number of patches, E=Embedding dimension. But this model input 2D tensor[B,E] to attention layer. Thus, Attention layer think this input has only one sample, not batch. So they convert [B,E] to [1,B,E] and run the attention operation (dot product). Therefore, when this model inference, they are not dot product about only one sample, dot product among the other samples in validation dataset. If the different label validation data within a batch exists during inference, their accuracy decreases rapidly.(originally, same label is clustered in validation dataset.ex) label1 ... label1, label2,...label2,... label n ,... , label n) Therefore, it cannot be said that the performance is correct,

Hey. I have tried to replace the keras Attention layer with Attention from maximal library. I am not 100% if it fixes the issue that you're describing, but I think the maximal Attention calculates the dot product for each vector individually? Could you please take a look at the attention function code from maximal library? The code is available on git hub: https://github.com/IvanBongiorni/maximal/blob/main/maximal/layers.py

**Comment by KrystianZielinski** [comment]
Now I am just wondering what's the idea? the attention mechanism should be calculated for each individual vector within a batch, that is coming from the FC layer, the 1x32 vector? and then calculate self attention only on that one vector, for each vector within the batch? or calculate only once self attention between all the vectors within the batch ? If so how is that supposed to work if there are images from different classes within a batch? Could you clarify the idea of the attention mechanics within this structure?

**Comment by KrystianZielinski** [comment]
I train the model on Shuffled train data set and the trained model gives GOOD! results on NON shuffled data sets. Also, I check the results on NON-shuffled same exact train dataset and the accuracy is increased by like 20%. So using the self attention layer implementation from maximal library works GOOD for training the model, but not for calculating the accuracy on shuffled datasets, probably the problem is that the images from multiple classes are within same batches. I wonder how could you modify the maximal self attention layer function to fix this issue. Also, I added a seed for weights initializations within the q, k and v dense layers for more stable training, as I noticed that sometimes the model would not start learning and would get stack at a very low accuracy during training. So generally the model works very well in FER task, but just gives misleading results for shuffled datasets.

**Comment by BerindeieAlexandru** [comment]
@KrystianZielinski hello, i also interested into fixing and testing this model, i tried with maximal layer but i think i got something wrong maybe when i tried to expand dims for new attention because model gets stuck during first 6 epochs, only training loss and accuracy increase, validation is stuck. I used a custom data generator and no shuffle for either sets. Would you be kind enough to share your fixed version, maybe i got something wrong at some point? Also i assume it would not affect model's performance when using it to new data? For example using in real time and there are quick emotion changes, would that cause the same issue for the attention layer? Thanks!

**Comment by KrystianZielinski** [comment]
> no shuffle for either sets

u don't shuffle training set for training? You should do that. It will only print lower accuracy on the training set but other than that the training procedure goes as it should. You can also verify it, and after training use eval on training set but not shuffled. The acc should be higher. 

> Also i assume it would not affect model's performance when using it to new data?

I haven't tested it on real-time camera input yet; I have only tested it on downloaded datasets. I also performed cross-database evaluation (trained on one dataset, test on different dataset) , and it generalizes well to images outside of the training data distribution.

**Comment by BerindeieAlexandru** [comment]
@KrystianZielinski i did with train only but this model behaves very very strange, maybe there are things that i do not yet understand. So when i did not use new layer just expand dims and shrink with original architecture i get around  45% at max then when i try with new layer (both with shuffle only for train, this with shuffle for validation too) I get around 30% at best. If no shuffle for validation model gets stuck at around 6 epochs and validation stays at exactly same same value just loss varies but that too very little difference. Maybe environment issue? I am using python 3.12 and latest TensorFlow version with CUDA. You did it on local or kaggle/colab?  I am interested on your take on fix, for me seems like something doesn't click (fine tuning part is also useless unfortunately, no significant improvement there max 1%).
